Added warning message for outdated (pre-QDM 5.3) measures

c-monkey opened the pull request:
JIRA: https://jira.mitre.org/browse/BONNIE-1074
TESTS: added unit tests in cql_logic_view_spec.js.coffee

I recommend comparing the test file with the `?w=1` url suffix because I had to indent most of the file to split into 2 test fixtures

If you wish to see this in a manual test, download a package (or use any you already have) prior to MAT 5.4 from: https://gitlab.mitre.org/bonnie/internal-documentation/wikis/cql-measures-for-testing, upload to Bonnie, and open the measure details screen.

holmesie commented on app/assets/javascripts/views/logic/cql_logic_view.js.coffee:
```diff
@@ -67,8 +68,20 @@ class Thorax.Views.CqlPopulationLogic extends Thorax.Views.BonnieView
     if Array.isArray @model.get('elm')
       _.each @model.get('elm'), (elm, elm_index) =>
+        _.each elm.library.usings.def, (id) =>
+          if (id.localIdentifier == "QDM")
```
It's not really Coffeescript style to use parens in If, Else, Unless, and Conditional Assignments, e.g. `if (stuff)`, but since I'm seeing it in other places in this file, I don't think it's an approval stopper.

c-monkey replied:
Nice catch. I was following the existing style as opposed to doing the right thing. Thanks!

c-monkey replied:
fixed this and some of the accompanying code

holmesie commented on app/assets/javascripts/views/logic/cql_logic_view.js.coffee:
```diff
@@ -67,8 +68,20 @@ class Thorax.Views.CqlPopulationLogic extends Thorax.Views.BonnieView
       _.each @model.get('elm'), (elm, elm_index) =>
+        _.each elm.library.usings.def, (id) =>
+          if (id.localIdentifier == "QDM")
+            versionParts = id.version.split('.')
```
Is it possible at any point for either `id` itself being null, or in this case, `version` being null (which causes the `.split` to raise an error)?

c-monkey replied:
I added a null check just in case.

holmesie commented on app/assets/javascripts/views/logic/cql_logic_view.js.coffee:
```diff
@@ -67,8 +68,20 @@ class Thorax.Views.CqlPopulationLogic extends Thorax.Views.BonnieView
+              major = parseInt(versionParts[0], 10)
+              minor = parseInt(versionParts[1], 10)
+              if (major < 5 || (major == 5 && minor < 3))
+                @hasOutdatedQDM = true
```
I think this all (previous 9 lines) can be simplified to:
```
version = parseFloat(id?.version)
unless isNaN(version)
   @hasOutdatedQDM = version < 5.3
```
parseFloat will ignore the second potential `.<number>`, so if we had `"5.0.2"`, it would return `5.0`.

NOTE: this would only make sense if we didn't ever care about patch numbers in QDM versions, which I am unsure of.

c-monkey replied:
Although extremely appealing, I think removing the flexibility isn't appropriate in the general case.

edeyoung commented on app/assets/javascripts/views/logic/cql_logic_view.js.coffee:
```diff
@@ -59,18 +59,31 @@ class Thorax.Views.CqlPopulationLogic extends Thorax.Views.BonnieView
+          if id?.localIdentifier == "QDM"
+            versionParts = id.version.split '.'
+            if versionParts.length <2
+              @hasOutdatedQDM = true
```
I doubt this will happen, but what if we go to QDM 5.3.1?

This would falsely fire.

c-monkey replied:
I have added a unit test to show QDM 5.3.1 is equal to or newer than QDM 5.3 and will now pass.

edeyoung commented on app/assets/javascripts/views/logic/cql_logic_view.js.coffee:
```diff
@@ -59,18 +59,31 @@ class Thorax.Views.CqlPopulationLogic extends Thorax.Views.BonnieView
+              major = parseInt versionParts[0], 10
+              minor = parseInt versionParts[1], 10
+              if major < 5 || (major == 5 && minor < 3)
+                @hasOutdatedQDM = true
```
What about something like this: 

```
Gem::Version.new('0.4.1') > Gem::Version.new('0.10.1')
```

from here: https://stackoverflow.com/questions/2051229/how-to-compare-versions-in-ruby

This would make it more generalized and we can then add the version we care about as a config variable (which we should do)

c-monkey replied:
I consent to this proposal.

c-monkey replied:
This would be perfect if this code were in Ruby. I'll figure out an equivalent for front-end.

edeyoung replied:
ha :) yes, thanks!

c-monkey replied:
The new CompareVersion class should handle this.

alexanderelliott121 commented:
I'm going to hold off on approval until the branch has been merged/rebased

holmesie commented on app/assets/javascripts/helpers/compare_version.js.coffee:
```diff
@@ -0,0 +1,15 @@
+class CompareVersion
```
I like this!

holmesie commented on config/bonnie.yml:
```diff
@@ -11,6 +11,8 @@ defaults: &defaults
     profile: Latest eCQM
   enable_client_error_email: true
+  # oldest version of QDM that can be loaded
+  support_qdm_version: 5.3
```
This will be nice to have

c-monkey commented:
Still investigating 1 travis failure. I fixed some BonnieRouter global variable assignment to work both in the App and in Jasmine tests, but am still investigating the other failure.

c-monkey commented:
This test failure is similar to one we fixed in the master branch. The test fixture is being changed by other tests execution. One can reproduce this test failure in the `cql4bonnie` branch in the web-teaspoon runner by running just the MeasureView suite. Regardless, this PR is going to be held up until I can fix that test. I also noticed, while debugging this, that the new suite I added `CompareVersion` always runs, for example when you select a suite to debug.

c-monkey commented:
rebased
